### PR TITLE
Zj/tips self serve s3

### PR DESCRIPTION
**Pull request by zacjones93**
[loom](https://www.loom.com/share/3f507a292a7d487aa18d612de2b703c5?sid=7b41b3de-ab50-4889-a633-ea1ade9268f5)

This adds tips self serve with the inngest flow that @joelhooks added recently. 

g served

### CODE REVIEW

File: src/app/tips/new/page.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import {redirect} from 'next/navigation'
+
+const NewTip = async () => {
+  const cookieStore = cookies()
+  const userToken = cookieStore?.get(ACCESS_TOKEN_KEY ?? '')?.value
+  const ability = await getAbilityFromToken(userToken)
+
+  if (!ability.can('upload', 'Video')) {
+    redirect('/')
+  }
```
Comment (zacjones93) [inline src/app/tips/new/page.tsx:14]: redirect to home if user doesn't have upload abilities

File: src/components/tips/tip-uploader.tsx [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+const TipCreationForm: React.FunctionComponent<
+  React.PropsWithChildren<any>
+> = () => {
+  const [s3Url, sets3Url] = React.useState<string>('')
+  const router = useRouter()
+  const form = useForm<z.infer<typeof NewTipFormSchema>>({
+    resolver: zodResolver(NewTipFormSchema),
+    defaultValues: {
+      title: '',
+      description: '',
```
Comment (zacjones93) [inline src/components/tips/tip-uploader.tsx:88]: I plan on adding more tip metadata like tags and body but this gets it started

File: src/components/tips/tip-uploader.tsx [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+            <FormItem>
+              <FormControl>
+                <Input type="hidden" {...field} />
+              </FormControl>
+              {form.formState.errors.s3Url && (
+                <p>{form.formState.errors.s3Url.message}</p>
+              )}
+            </FormItem>
+          )}
+        />
```
Comment (zacjones93) [inline src/components/tips/tip-uploader.tsx:200]: the error here says you need atleast 4 characters which is confusing, need to update with a custom error message that says please upload a video

File: src/components/tips/tip-uploader.tsx [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+          }}
+        >
+          Submit Tip
+        </Button>
+      </form>
+    </Form>
+  )
+}
+
+function TipUploader({sets3Url}: {sets3Url: (value: string) => void}) {
```
Comment (zacjones93) [inline src/components/tips/tip-uploader.tsx:221]: there was a lot of logic here which is the original S3 uploader so put it in it's own component

File: src/inngest/functions/mux/add-srt-to-mux-asset.ts [hunk 5]
```diff
@@ -21,7 +22,7 @@ export const addSrtToMuxAsset = inngest.createFunction(
       'get the video resource from Sanity',
       async () => {
         const resourceTemp = VideoResourceSchema.safeParse(
-          await sanityQuery(
+          await sanityWriteClient.fetch(
```
Comment (zacjones93) [inline src/inngest/functions/mux/add-srt-to-mux-asset.ts:25]: using the sanity client

File: src/inngest/functions/mux/add-srt-to-mux-asset.ts [hunk 6]
```diff
@@ -31,7 +32,7 @@ export const addSrtToMuxAsset = inngest.createFunction(
 
     if (videoResource) {
       const muxAsset = await step.run('get the mux asset', async () => {
-        const assetId = videoResource.muxAssetId
+        const assetId = videoResource.muxAsset.muxAssetId
```
Comment (zacjones93) [inline src/inngest/functions/mux/add-srt-to-mux-asset.ts:35]: adjusting for egghead data sanity structure

File: src/inngest/functions/tip-video-uploaded.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
+import {inngest} from '@/inngest/inngest.server'
+import {TIP_VIDEO_UPLOADED_EVENT} from '@/inngest/events/tips'
+import {sanityMutation, sanityQuery} from '@/utils/sanity.fetch.only.server'
+import {sanityWriteClient} from '@/utils/sanity-server'
+import {getMuxOptions} from '@/lib/get-mux-options'
+import {v4} from 'uuid'
+import {orderDeepgramTranscript} from '@/lib/deepgram-order-transcript'
+import {createMuxAsset} from '../../lib/mux'
+
+export const tipVideoUploaded = inngest.createFunction(
```
Comment (zacjones93) [inline src/inngest/functions/tip-video-uploaded.ts:10]: created a `tipVideoUploaded` function

File: src/lib/deepgram-order-transcript.ts [hunk 8]
```diff
@@ -17,6 +17,7 @@ export async function orderDeepgramTranscript({
   })
 
   // just weird URL differences between dev and prod
+  // replace with ngrok tunnel in dev
   const callbackBase = process.env.NEXT_PUBLIC_DEPLOYMENT_URL
```
Comment (zacjones93) [inline src/lib/deepgram-order-transcript.ts:21]: this took me a bit to figure out...

File: src/server/routers/tips.ts [hunk 9]
```diff
@@ -19,24 +21,21 @@ export const tipsRouter = router({
...
-        // create the video resource object in Sanity
+      if (ability.can('upload', 'Video')) {
         const newVideoResource = await sanityWriteClient.create({
           _id: `videoResource-${v4()}`,
           _type: 'videoResource',
           state: 'new',
-          title: input.fileName,
-          originalMediaUrl: input.s3Url,
+          filename: input.fileName,
+          originalVideoUrl: input.s3Url,
```
Comment (zacjones93) [inline src/server/routers/tips.ts:37]: slightly different names for the original video url

File: src/server/routers/tips.ts [hunk 10]
```diff
@@ -46,16 +45,34 @@ export const tipsRouter = router({
...
               // since title is unique, we can use it as the slug with a random string
               current: `${slugify(input.title)}~${nanoid()}`,
             },
+            collaborators: [
+              {
+                _key: v4(),
+                _type: 'reference',
+                _ref: sanityInstructor._id,
+              },
+            ],
```
Comment (zacjones93) [inline src/server/routers/tips.ts:75]: add the instructor as well in the tip that gets created

File: src/components/tips/tip-uploader.tsx [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+const formatString = (str: string) => {
+  return toChicagoTitleCase(removeFileExtension(str))
+}
+
+const NewTipFormSchema = z.object({
+  title: z.string().min(10).max(90),
+  description: z.string().min(2).max(240),
+  s3Url: z.string().min(4),
```
Comment (joelhooks) [inline src/components/tips/tip-uploader.tsx:76]:
```suggestion
  s3Url: z.string().trim().url({ message: 'Please Upload a Video' }),
```

File: src/components/tips/tip-uploader.tsx (hunk 2 above)
Comment (joelhooks) [reply src/components/tips/tip-uploader.tsx:88]: I think the initial draft creation form can be very simple.

File: src/components/tips/tip-uploader.tsx (hunk 3 above)
Comment (joelhooks) [reply src/components/tips/tip-uploader.tsx:200]: `z.string().trim().url({ message: 'Please Upload a Video' })`

added this to the schema

File: src/inngest/events/sanity-post.ts [hunk 12]
```diff
@@ -0,0 +1,18 @@
+import {z} from 'zod'
+
+export const POST_CREATION_REQUESTED_EVENT = 'video/post-creation-requested'
```
Comment (joelhooks) [inline src/inngest/events/sanity-post.ts:3]: this whole file can go away?

**Review by joelhooks — APPROVED** [review]

File: src/inngest/events/sanity-post.ts (hunk 12 above)
Comment (zacjones93) [reply src/inngest/events/sanity-post.ts:3]: yeah that isn't being used

**Pull request merged**